[N] Line chart - Add styling & typings

IIsaku opened the pull request:
Port styling and typings of the new native Line chart widget from Atlas 3.

IIsaku committed:
Add styling & typings

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+        fontFamily: font.family,
+        fontSize: font.sizeSmall,
+    },
+    chart: {
```
I don't think we need this prop as we already have `container`

IIsaku replied:
`container` view can contain `errorMessage` text or `chart` view. Not sure if combining them is ideal...

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+    chart: {
+        // All ViewStyle properties are allowed
+    },
+    gridAndLabelsRow: {
+        // All ViewStyle properties are allowed
+    },
+    gridRow: {
+        // All ViewStyle properties are allowed
+    },
+    gridWrapper: {
+        // All ViewStyle properties are allowed
+    },
```
I'm not sure what the difference is between these 3 props. Do we really need them exposed? If so, how do we make it more clear what they represent?

cjfhodges replied:
Agree quite confusing, what represents what. Do we need three viewstyles? Is this used for alignment?

IIsaku replied:
Removed the view styles, since you don't need to style them. The only ones left are `container` and `chart`.

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+        // All ViewStyle properties are allowed
+    },
+    grid: {
+        padding: { top: 8, right: 8, bottom: 32, left: 32 },
```
Imo this can be changed to `paddingTop`, `paddingBottom` etc. Because these are real native styling props.

cjfhodges replied:
yes use the padding: variables declared by Atlas where possible.

IIsaku replied:
Yes, will changed that. Using Atlas variables here isn't wise, since this is about the padding inside the svg canvas. We actually don't know how much padding we need to make long tick value labels on the axes fully visible. Jordan and I are going to brainstorm about a solution for this.

cjfhodges replied:
is it best to add this padding then to core styling of the widget?

wegiswes replied:
What is the outcome of the brainstorm? 
Is padding dynamic here? If not, I think using variables is a good idea

IIsaku replied:
We shouldn't reuse spacing variables here, because if the values are changed by users, the axis tick value labels could become invisible creating confusion. We will look for padding calculation in the future, so that devs don't need to set this anymore to make tick value labels visible.

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+            tickLabels: { fontFamily: font.family, fontSize: font.sizeSmall },
+        },
+    },
+    xAxisLabel: {
```
Imo it makes more sense to add this inside `xAxis` & `yAxis`. Because we have 2 different labels my suggestion for the names are `innerLabel` and `outerLabel`

cjfhodges replied:
Agree, whats the difference between `innerLabel` and `outerLabel` ? Can't we just have one prop `axisLabel`

IIsaku replied:
Moved this label inside the xAxis style property

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+        fontFamily: font.family,
+        fontSize: font.sizeSmall,
+    },
+    yAxisLabel: {
```
Same as above

IIsaku replied:
Moved this label inside the yAxis style property.

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+        fontFamily: font.family,
+        fontSize: font.sizeSmall,
+    },
+    lineStyles: {},
```
What is expected in this object?

IIsaku replied:
This will be explained in the docs as the styles are quite custom. It would be a lot to write as just a comment. Or should we add it anyway?

wegiswes replied:
I think it's a good idea to either add a comment explaining what is expected, or link to the right docs

IIsaku replied:
Added a comment with allowed properties.

wegiswes commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+            padding: 0,
+            paddingRight: 16,
+        },
+        indicator: {
+            // All ViewStyle properties are allowed
+            marginRight: 8,
+        },
+        label: {
+            // All TextStyle properties are allowed
+            fontFamily: font.family,
+            fontSize: font.sizeSmall,
+        },
```
If indicator and label are inside `item`, I would move these props there

cjfhodges replied:
`marginVertical` and `marginHorizontal` should also be based off atlas variables

cjfhodges replied:
My thoughts on the styling of charting was that we define a new set of styles for charting
```
export const chartBrand = {
  gridStrokeColor: "#CED0D3",
  labelColor: "#0A1325",
  tickColor: "transparent",
  colorScale: "#264AE5", "#03096C", "#E6930E", "#3CB33D"
};
```
is a similar approach to how excel, powerbi and tableau have like a charting theme.

IIsaku replied:
This idea really sounds as what @wesleysieses was talking about. Not exposing the full library styling, but just define a small set of self defined styles.

cjfhodges commented on styles/native/ts/core/widgets/linechart.ts:
```diff
@@ -0,0 +1,92 @@
+    grid: {
+        padding: { top: 8, right: 8, bottom: 32, left: 32 },
+        xAxis: {
+            grid: { stroke: "#CED0D3" },
```
stoke color should be based off a atlas variable

IIsaku replied:
Done. It now `lineColor` inside the `grid` style property.

IIsaku committed:
Refactor line chart native styles

- Use new simple styles

IIsaku committed:
feat(line-chart-native): change styles

- Rename color to lineColor for line colors
- Add color style for axis tick labels
- Reuse more variables inside to support dark mode as well

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+    // All ViewStyle properties are allowed
+    },
+    grid: {
+        lineColor: border.color,
```
This object is missing a comment

IIsaku replied:
Added a comment with allowed properties.

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+        paddingTop: 8
+    },
+    xAxis: {
+        color: font.color,
```
This object is missing a comment

IIsaku replied:
Added a comment with allowed properties.

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+            fontSize: font.sizeSmall,
+            fontWeight: font.weightNormal
+        },
+        lineColor: border.color
```
I would move `lineColor` on top of `label`

IIsaku replied:
Everything is in alphabetical order. To keep it consistent I would like to keep it here.

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+            fontSize: font.sizeSmall,
+            fontWeight: font.weightNormal
+        },
+        lineColor: border.color
```
I would move `lineColor` on top of `label`

IIsaku replied:
Everything is in alphabetical order. To keep it consistent I would like to keep it here.

wegiswes replied:
It's only this subobject which is on alphabetical order? All the other line chart styles are not. I would put a clear structure above alphabetical order

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+            color: font.color,
+            alignSelf: "center",
+            marginHorizontal: 0,
+            marginVertical: 8,
```
We can use the `spacing` variable here

IIsaku replied:
Done

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+            // All TextStyle properties are allowed & relativePositionGrid property ("top" or "left")
+            color: font.color,
+            marginHorizontal: 0,
+            marginVertical: 8,
```
We can use the `spacing` variable here

IIsaku replied:
Done

wegiswes commented on styles/native/js/core/widgets/linechart.js:
```diff
@@ -0,0 +1,94 @@
+            // All ViewStyle properties are allowed
+            justifyContent: "flex-start",
+            marginHorizontal: 0,
+            marginVertical: 8
```
We can use the `spacing` variable here, also in `item` & in `indicator`

IIsaku replied:
Done

IIsaku committed:
Improve line chart native styles

- Add descriptions for allowed properties
- Use spacing variables

IIsaku committed:
Use spacing variable for axis labels

IIsaku committed:
Minor line chart style adjustments

IIsaku committed:
feat(line-chart-native): change styling interface

- Rename 'lineStyles' to 'lines', because we refer to elements in the root of style objects for widgets.
- Add style property 'lineColorPalette' and 'customLineStyles' to 'lines' style property.
- 'customLineStyles' contains self named line style properties whose values adhere to the line style interface.
- Change 'lineColorPalette' to a string, because array style properties are concatenated by the Mendix native client instead of being overridden.

IIsaku committed:
feat(line-chart-native): change width to lineWidth

IIsaku committed:
feat(line-chart-native): add design property

- Add chart size design property to easily make chart grids visible inside a scroll container
